### ISSUE TITLE
Maybe we should consider avoiding dependence on nipy

### ISSUE THREAD

**Issue by ganow** [KamitaniLab/bdpy#68]
While creating a [PR for new dependency management, including extra dependencies](https://github.com/KamitaniLab/bdpy/pull/67), I noticed that the [`nipy`](https://github.com/nipy/nipy) installation fails with Python>=3.10.

c.f. https://github.com/KamitaniLab/bdpy/pull/67#issuecomment-1600789516

A similar issue has already [been discussed on the `nipy` GitHub issue](https://github.com/nipy/nipy/issues/494), but unfortunately according to this discussion, nipy is ["essentially unmaintained."](https://github.com/nipy/nipy/issues/494#issuecomment-1232928167) In fact, it appears that the developers of the other packages that initiated this discussion [have decided to replace the nipy-dependent portions of their codebase](https://github.com/aramis-lab/clinica/issues/734).

As far as I know at this time, bdpy currently uses at least the following features of nipy:

- `nipy.load_image`
- `nipy.modalities.fmri.experimental_paradigm.{BlockParadigm,EventRelatedParadigm}`

I don't know at this time how many resources would be needed to replace these codebases.

**Comment by ganow** [comment]
## Why does `nipy` installation via `pyproject.toml` fail?

First, we assume the following structure for `pyproject.toml`:

```toml
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"
...
dependencies = ["numpy<1.24", "nipy"]
```

`nipy` is a tool that depends on `numpy`. If you try to install it in the project folder with Python 3.10 or later using the command:

```bash
$ pip install .
...
Collecting nipy
  Using cached nipy-0.5.0.tar.gz (2.4 MB)
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error

  × python setup.py egg_info did not run successfully.
 ...
      ModuleNotFoundError: No module named 'numpy'
```

The installation fails. The cause of this error originates from the following procedure:
1. Since `nipy` is not provided in wheel format for Python 3.10 or later, the build tool attempts to build from source code in sdist format.
2. Because there's a statement `from numpy.distutils.misc_util import appendpath` in `setup.py`, an exception is thrown if `numpy` is not installed in the runtime **before** this line is evaluated.

This issue can be circumvented by first installing `numpy` only. In other words, the following installation process, for example, would not cause a problem:

```bash
$ pip install 'numpy<1.24'
$ pip install .  # includes installation of nipy
```

The problem with this approach is that it's not possible to declaratively describe the installation procedure. Also, as mentioned in the original issue, manual building in wheel format, as shown below, will pass without problems as long as `numpy` is installed beforehand.

```bash
$ pip install build
$ pip install 'numpy<1.23'
$ git clone https://github.com/nipy/nipy.git && cd nipy
$ python -m build
```

**Comment by ganow** [comment]
## Update on 6/26

### TL;DR

If we wait about several months to a year, the problem will be solved 🙌.

### Detail

I found that a `pyproject.toml` was added to `nipy` about ~2mo ago. https://github.com/nipy/nipy/commit/adccc98f365b01123a6aaa59e3f4adeaae815170

The following procedures work on the current nipy master branch.

```bash
$ git clone https://github.com/nipy/nipy.git && cd nipy
$ pip install .
```

When the next version of nipy is released, it will likely be possible to put a declarative dependency on nipy in our pyproject.toml.

**Comment by ganow** [comment]
The following can solve the problem tentatively:

```toml
[project.optional-dependencies]
mri = [
    "nipy ; python_version < '3.10'",
    "nipy @ git+https://github.com/nipy/nipy.git@7eede02 ; python_version >= '3.10'"  # build from recent git commit
]
```
